pallets/proxy-bonding/src/lib.rs

```diff
@@ -0,0 +1,197 @@
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+// Needed due to empty sections raising the warning
+#![allow(unreachable_patterns)]
+pub use pallet::*;
+
+mod functions;
+
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use frame_support::{
+		pallet_prelude::{Weight, *},
+		traits::{
+			fungible,
+			fungible::{Mutate, MutateHold},
+			fungibles,
+			fungibles::{Inspect as FungiblesInspect, Mutate as FungiblesMutate},
+			tokens::{Precision, Preservation},
+		},
+	};
+	use frame_system::pallet_prelude::*;
+	use polimec_common::ProvideAssetPrice;
+	use sp_runtime::{
+		traits::{AccountIdConversion},
+		Perbill, TypeId,
+	};
+
+	pub type AssetId = u32;
+	pub type BalanceOf<T> = <<T as Config>::BondingToken as fungible::Inspect<AccountIdOf<T>>>::Balance;
+	pub type AccountIdOf<T> = <T as frame_system::Config>::AccountId;
+	pub type HoldReasonOf<T> = <<T as Config>::BondingToken as fungible::InspectHold<AccountIdOf<T>>>::Reason;
+
+	/// Configure the pallet by specifying the parameters and types on which it depends.
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// Because this pallet emits events, it depends on the runtime's definition of an event.
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		type RuntimeHoldReason: IsType<HoldReasonOf<Self>> + Parameter + MaxEncodedLen;
+
+		/// The pallet giving access to the bonding token
+		type BondingToken: fungible::Inspect<Self::AccountId>
+			+ fungible::Mutate<Self::AccountId>
+			+ fungible::MutateHold<Self::AccountId>;
+
+		type BondingTokenDecimals: Get<u8>;
+		type UsdDecimals: Get<u8>;
+		type BondingTokenId: Get<AssetId>;
+
+		/// The pallet giving access to fee-paying assets, like USDT
+		type FeeToken: fungibles::Inspect<Self::AccountId, Balance = BalanceOf<Self>, AssetId = AssetId>
+			+ fungibles::Mutate<Self::AccountId, Balance = BalanceOf<Self>, AssetId = AssetId>
+			+ fungibles::metadata::Inspect<Self::AccountId, Balance = BalanceOf<Self>, AssetId = AssetId>;
+
+		type FeePercentage: Get<Perbill>;
+
+		/// Method to get the price of an asset like USDT or PLMC. Likely to come from an oracle
+		type PriceProvider: ProvideAssetPrice<AssetId = u32>;
+
+		/// The account holding the tokens to be bonded. Normally the treasury
+		type Treasury: Get<Self::AccountId>;
+
+		/// The account receiving the fees
+		type FeeRecipient: Get<Self::AccountId>;
+
+		/// The id type that can generate sub-accounts
+		type Id: Encode + Decode + TypeId;
+
+		/// The root id used to derive sub-accounts. These sub-accounts will be used to bond the tokens
+		type RootId: Get<Self::Id>;
+	}
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	#[derive(Clone, Encode, Decode, Eq, PartialEq, RuntimeDebug, MaxEncodedLen, TypeInfo)]
+	pub enum ReleaseType<BlockNumber> {
```

Comment:
Usually for the reasons we use the `pallet::composite_enum` [attribute]([url](https://paritytech.github.io/polkadot-sdk/master/frame_support/pallet_macros/attr.composite_enum.html)). Do we have to use “manual” derive in this case because of the generic parameter?

Reply:
we can no longer have inner values inside the variants of a composite enum, so I cannot add the `Locked(BlockNumber)` as part of the hold reason. Also this is not really a hold reason, and we cannot have custom composite enums, only HoldReason, FreezeReason, etc